File: backend/app/models/punch_event.py
import uuid
from datetime import datetime
from sqlalchemy import String, Float, DateTime, Boolean, Integer, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column
from pydantic import BaseModel, Field

from app.models.attendance import Base


class PunchEvent(Base):
    __tablename__ = "punch_events"

    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    worker_id: Mapped[str] = mapped_column(String(64), ForeignKey("workers.id"), index=True)
    type: Mapped[str] = mapped_column(String(8))  # 'in' or 'out'
    timestamp: Mapped[datetime] = mapped_column(DateTime(timezone=True), index=True)
    gps_lat: Mapped[float | None] = mapped_column(Float, nullable=True)
    gps_lon: Mapped[float | None] = mapped_column(Float, nullable=True)
    gps_accuracy: Mapped[float | None] = mapped_column(Float, nullable=True)
    face_match_score: Mapped[float | None] = mapped_column(Float, nullable=True)
    liveness_passed: Mapped[bool] = mapped_column(Boolean, default=True)
    device_id: Mapped[str | None] = mapped_column(String(128), nullable=True)
    sync_attempts: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default="now()")


# ---------- Pydantic ----------

class PunchEventIn(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    type: str = Field(pattern="^(in|out)$")
    timestamp: datetime
    gps_lat: float | None = None
    gps_lon: float | None = None
    gps_accuracy: float | None = None
    face_match_score: float | None = None
    liveness_passed: bool = True
    device_id: str | None = None


class PunchEventBatchIn(BaseModel):
    events: list[PunchEventIn]


class PunchEventBatchOut(BaseModel):
    accepted: list[str]
    rejected: list[str]


class PunchEventOut(BaseModel):
    id: str
    worker_id: str
    type: str
    timestamp: datetime
    gps_lat: float | None
    gps_lon: float | None
    face_match_score: float | None
    liveness_passed: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class AttendanceSummary(BaseModel):
    """Per-day rollup for calendar view."""
    date: str  # YYYY-MM-DD
    punch_in: datetime | None
    punch_out: datetime | None
    duration_minutes: int | None
    status: str  # 'full' | 'partial' | 'absent'

File: ios/lib/models/punch_event.dart
class PunchEvent {
  final String id;
  final String type;
  final DateTime timestamp;
  final double? gpsLat;
  final double? gpsLon;
  final double? gpsAccuracy;
  final double? faceMatchScore;
  final bool livenessPassed;
  final String deviceId;
  final int synced;

  PunchEvent({
    required this.id,
    required this.type,
    required this.timestamp,
    this.gpsLat,
    this.gpsLon,
    this.gpsAccuracy,
    this.faceMatchScore,
    required this.livenessPassed,
    required this.deviceId,
    required this.synced,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type,
        'timestamp': timestamp.toIso8601String(),
        'gpsLat': gpsLat,
        'gpsLon': gpsLon,
        'gpsAccuracy': gpsAccuracy,
        'faceMatchScore': faceMatchScore,
        'livenessPassed': livenessPassed,
        'deviceId': deviceId,
        'synced': synced,
      };

  factory PunchEvent.fromJson(Map<String, dynamic> json) {
    return PunchEvent(
      id: json['id'] as String,
      type: json['type'] as String,
      timestamp: DateTime.parse(json['timestamp'] as String),
      gpsLat: (json['gpsLat'] as num?)?.toDouble(),
      gpsLon: (json['gpsLon'] as num?)?.toDouble(),
      gpsAccuracy: (json['gpsAccuracy'] as num?)?.toDouble(),
      faceMatchScore: (json['faceMatchScore'] as num?)?.toDouble(),
      livenessPassed: json['livenessPassed'] as bool,
      deviceId: json['deviceId'] as String,
      synced: json['synced'] as int,
    );
  }

  Map<String, dynamic> toMap() => {
        'id': id,
        'type': type,
        'timestamp': timestamp.toIso8601String(),
        'gps_lat': gpsLat,
        'gps_lon': gpsLon,
        'gps_accuracy': gpsAccuracy,
        'face_match_score': faceMatchScore,
        'liveness_passed': livenessPassed ? 1 : 0,
        'device_id': deviceId,
        'synced': synced,
      };

  factory PunchEvent.fromMap(Map<String, dynamic> map) {
    return PunchEvent(
      id: map['id'] as String,
      type: map['type'] as String,
      timestamp: DateTime.parse(map['timestamp'] as String),
      gpsLat: (map['gps_lat'] as num?)?.toDouble(),
      gpsLon: (map['gps_lon'] as num?)?.toDouble(),
      gpsAccuracy: (map['gps_accuracy'] as num?)?.toDouble(),
      faceMatchScore: (map['face_match_score'] as num?)?.toDouble(),
      livenessPassed: (map['liveness_passed'] as int) == 1,
      deviceId: map['device_id'] as String,
      synced: map['synced'] as int,
    );
  }
}
